frame/pools/src/lib.rs

```diff
@@ -0,0 +1,1135 @@
+				SubPoolsStorage::<T>::get(&delegator.pool).ok_or(Error::<T>::SubPoolsNotFound)?;
+
+			let balance_to_unbond = if let Some(pool) = sub_pools.with_era.get_mut(&unbonding_era) {
+				let balance_to_unbond = pool.balance_to_unbond(delegator.points);
+				pool.points = pool.points.saturating_sub(delegator.points);
+				pool.balance = pool.balance.saturating_sub(balance_to_unbond);
+
+				balance_to_unbond
+			} else {
+				// A pool does not belong to this era, so it must have been merged to the era-less
+				// pool.
+				let balance_to_unbond = sub_pools.no_era.balance_to_unbond(delegator.points);
+				sub_pools.no_era.points = sub_pools.no_era.points.saturating_sub(delegator.points);
+				sub_pools.no_era.balance =
+					sub_pools.no_era.balance.saturating_sub(balance_to_unbond);
+
+				balance_to_unbond
+			};
+
+			T::StakingInterface::withdraw_unbonded(delegator.pool.clone(), num_slashing_spans)?;
+
+			T::Currency::transfer(
+				&delegator.pool,
+				&who,
+				balance_to_unbond,
+				ExistenceRequirement::AllowDeath,
+			)
+			.map_err(|e| {
+				log::warn!("system logic error: pool could not withdraw_unbonded due to lack of free balance");
+				e
+			})?;
+
+			SubPoolsStorage::<T>::insert(&delegator.pool, sub_pools);
+			Delegators::<T>::remove(&who);
+
+			Self::deposit_event(Event::<T>::Withdrawn {
+				delegator: who,
+				pool: delegator.pool,
+				amount: balance_to_unbond,
+			});
+
+			Ok(())
+		}
+
+		// Create a pool.
+		//
+		// * `targets`: _Stash_ addresses of the validators to nominate
+		// * `amount`: Balance to delegate to the pool. Must meet the minimum bond.
+		// * `index`: Disambiguation index for seeding account generation. Likely only useful when
+		//   creating multiple pools in the same extrinsic.
+		#[pallet::weight(666)]
+		pub fn create(
+			origin: OriginFor<T>,
+			targets: Vec<<T::Lookup as StaticLookup>::Source>,
+			amount: BalanceOf<T>,
+			index: u16,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(amount >= T::StakingInterface::minimum_bond(), Error::<T>::MinimumBondNotMet);
+
+			let (stash, reward_dest) = Self::create_accounts(index);
+
+			ensure!(!BondedPools::<T>::contains_key(&stash), Error::<T>::IdInUse);
+
+			T::StakingInterface::bond_checks(&stash, &stash, amount, &reward_dest)?;
+			let (stash, targets) = T::StakingInterface::nominate_checks(&stash, targets)?;
+
+			let mut bonded_pool = BondedPool::<T> { points: Zero::zero() };
+
+			// We must calculate the points to issue *before* we bond who's funds, else
+			// points:balance ratio will be wrong.
+			let points_to_issue = bonded_pool.points_to_issue(&stash, amount);
+			bonded_pool.points = points_to_issue;
+
+			T::Currency::transfer(&who, &stash, amount, ExistenceRequirement::AllowDeath)?;
+
+			T::StakingInterface::bond(
+				stash.clone(),
+				// We make the stash and controller the same for simplicity
+				stash.clone(),
+				amount,
+				reward_dest.clone(),
+			)
+			.map_err(|e| {
```

Comment:
```suggestion
			.defensive_map_err(|e| {
```

because we already did the `ensure_can_bond` before.

Reply:
(or move to transactional as spoke offline)

Reply:
d0ebdc1458c45bddfa241e4bf89f74d60a65a635